server/src/test/java/io/spine/examples/shareaware/server/e2e/SharePurchaseTest.java

```diff
@@ -0,0 +1,139 @@
+import io.spine.examples.shareaware.server.e2e.given.E2EUser;
+import io.spine.examples.shareaware.server.e2e.given.SubscriptionOutcome;
+import io.spine.examples.shareaware.server.e2e.given.WithClient;
+import io.spine.examples.shareaware.share.Share;
+import io.spine.examples.shareaware.wallet.WalletBalance;
+import io.spine.examples.shareaware.wallet.rejection.Rejections.InsufficientFunds;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.truth.extensions.proto.ProtoTruth.assertThat;
+import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
+import static io.spine.examples.shareaware.given.GivenMoney.usd;
+import static io.spine.examples.shareaware.server.e2e.given.E2EUserTestEnv.purchaseSharesFor;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.balanceAfterTeslaPurchase;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.chooseTeslaShareFrom;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.insufficientFundsAfter;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.investmentAfterTeslaPurchase;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.walletBalanceWith;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.zeroWalletBalance;
+import static java.time.Duration.ofMillis;
+
+/**
+ * End-to-end test that describes such a scenario:
+ * <ol>
+ *     <li>The user attempts to purchase shares with no money in the wallet.</li>
+ *     <li>The user replenishes his wallet for 500 dollars.</li>
+ *     <li>The user successfully purchases one 'Tesla' share.</li>
+ *     <li>The user withdraws all his money from the wallet.</li>
+ * </ol>
+ */
+final class SharePurchaseTest extends WithClient {
+
+    @Test
+    @DisplayName("User should purchase one tesla share and withdraw all the money after this")
+    void test() throws ExecutionException, InterruptedException {
+        SharePurchaseUser user = new SharePurchaseUser(client());
+
+        WalletBalance initialBalance = user.signsUp();
+        sleepUninterruptibly(ofMillis(1500));
+        List<Share> shares = user.looksAtShares();
+        Share tesla = chooseTeslaShareFrom(shares);
+
+        WalletBalance balanceAfterFailedPurchase = user.attemptsToPurchaseShare(tesla);
+        assertThat(balanceAfterFailedPurchase).isEqualTo(initialBalance);
+
+        WalletBalance balanceAfterReplenishment = user.replenishesWalletFor(usd(500));
+        WalletBalance expectedBalanceAfterReplenishment =
+                walletBalanceWith(usd(500), user.walletId());
+        assertThat(balanceAfterReplenishment).isEqualTo(expectedBalanceAfterReplenishment);
```

Comment:
Also, how about we introduce `ClientWallet` as a client-level abstraction reflecting the current state of user's wallet? It would subscribe to all the changes made to the wallet. And you'll be able to use its state to verify changes made.